src/ott/solvers/nn/neuraldual.py

```diff
@@ -219,65 +408,93 @@ def train_neuraldual(
 
     return {"train_logs": train_logs, "valid_logs": valid_logs}
 
-  def get_step_fn(self, train: bool, to_optimize: Literal["f", "g"] = "g"):
-    """Create a one-step training and evaluation function."""
+  def get_step_fn(
+      self, train: bool, to_optimize: Literal["f", "g", "parallel"]
+  ):
+    """Create a parallel training and evaluation function."""
 
     def loss_fn(params_f, params_g, f, g, batch):
-      """Loss function for potential f."""
+      """Loss function for both potentials."""
       # get two distributions
       source, target = batch["source"], batch["target"]
 
-      # get loss terms of kantorovich dual
-      f_t = f({"params": params_f}, batch["target"])
-
-      grad_g_s = jax.vmap(
-          lambda x: jax.grad(g, argnums=1)({
-              "params": params_g
-          }, x)
-      )(
-          batch["source"]
-      )
-
-      f_grad_g_s = f({"params": params_f}, grad_g_s)
-
-      s_dot_grad_g_s = jnp.sum(source * grad_g_s, axis=1)
+      if self.g_returns_potential:
+        g_grad = jax.vmap(
```

Comment:
For clarify, I think this could be a method of `PotentialMixin` as
```python
def grad(self, params) -> Callable[[jnp.ndarray], ndarray]:
   if self.is_potential:
       return jax.vmap(lambda y: jax.grad(self, ...))
   return self  # __call__ already defined a gradient
```

Reply:
That is nice! And the potential's objective value can be put into a similar interface instead of inlined as I've done in `to_dual_potentials`.